File: examples/go/clean-api-go/application/commands/register_sale.go
package commands

import (
	"context"
	"errors"

	bberrors "github.com/FreakJazz/backbone/backbone-go/errors"
	"github.com/FreakJazz/backbone/backbone-go/infrastructure/logging"
	bbex "github.com/FreakJazz/backbone/backbone-go/interfaces/responses"
	"github.com/freakjazz/clean-api-go/domain/entities"
	"github.com/freakjazz/clean-api-go/domain/repositories"
)

type RegisterSaleCommand struct {
	ProductID string
	Quantity  int
}

// RegisterSaleCommandHandler coordinates two independent stores for one
// business action: PostgreSQL owns the authoritative stock count, MongoDB
// owns the append-only sales log. There is no distributed transaction here
// — stock is decremented first (the step that can legitimately fail, e.g.
// insufficient stock) and the sale is recorded only after that succeeds. If
// the Mongo insert itself fails, stock has already moved and this returns
// an error without rolling it back; a production system would reconcile
// that gap with an outbox/saga instead of a bare two-step call.
type RegisterSaleCommandHandler struct {
	products repositories.IProductRepository
	sales    repositories.ISaleRepository
	logger   *logging.EnhancedLogger
}

func NewRegisterSaleCommandHandler(products repositories.IProductRepository, sales repositories.ISaleRepository) *RegisterSaleCommandHandler {
	return &RegisterSaleCommandHandler{
		products: products,
		sales:    sales,
		logger:   logging.NewEnhancedLogger("clean-api-go").WithLayer("application").WithComponent("RegisterSaleCommandHandler").WithMethod("Handle"),
	}
}

func (h *RegisterSaleCommandHandler) Handle(ctx context.Context, cmd RegisterSaleCommand) (string, *bbex.ErrorResponse) {
	log := h.logger

	if cmd.Quantity <= 0 {
		e := bbex.ErrorResponseBuilder.ValidationError("quantity must be greater than 0",
			bbex.ErrorOpts{Code: bberrors.IfcInvalidRequestBody.Int()})
		return "", &e
	}

	product, err := h.products.FindByID(ctx, cmd.ProductID)
	if err != nil || product == nil {
		e := bbex.ErrorResponseBuilder.NotFound("product not found")
		return "", &e
	}

	if _, err := h.products.AdjustStock(ctx, cmd.ProductID, -cmd.Quantity); err != nil {
		if errors.Is(err, repositories.ErrInsufficientStock) {
			e := bbex.ErrorResponseBuilder.Conflict("insufficient stock for this sale",
				bbex.ErrorOpts{Code: bberrors.AppConflict.Int()})
			return "", &e
		}
		log.ErrorWithCode("AdjustStock failed", bberrors.InfraDBFailure.Int(), map[string]interface{}{"error": err.Error()})
		e := bbex.ErrorResponseBuilder.InternalServerError(err.Error())
		return "", &e
	}

	sale := entities.NewSale(cmd.ProductID, cmd.Quantity, product.Price)
	saved, err := h.sales.Save(ctx, sale)
	if err != nil {
		// Stock was already decremented in Postgres — flagged here rather
		// than silently swallowed, since this is exactly the inconsistency
		// window a saga/outbox pattern exists to close.
		log.ErrorWithCode("Sale record failed after stock was decremented — data inconsistency", bberrors.InfraDBFailure.Int(), map[string]interface{}{
			"product_id": cmd.ProductID, "quantity": cmd.Quantity, "error": err.Error(),
		})
		e := bbex.ErrorResponseBuilder.InternalServerError("sale could not be recorded; stock was already adjusted")
		return "", &e
	}

	log.Info("Sale registered", map[string]interface{}{"sale_id": saved.ID, "product_id": cmd.ProductID, "quantity": cmd.Quantity})
	return saved.ID, nil
}

File: examples/python/clean_api_python/application/commands/register_sale.py
from dataclasses import dataclass

from backbone.errors import ErrorCodes
from backbone.application.exceptions import ValidationException, ResourceNotFoundException, ResourceConflictException
from backbone.infrastructure.exceptions import DatabaseException
from backbone.infrastructure.logging import LoggerFactory

from domain.entities.sale import Sale
from domain.repositories.product_repository import IProductRepository, InsufficientStockError
from domain.repositories.sale_repository import ISaleRepository


@dataclass
class RegisterSaleCommand:
    product_id: str
    quantity: int


class RegisterSaleCommandHandler:
    """Coordinates two independent stores for one business action: PostgreSQL
    owns the authoritative stock count, MongoDB owns the append-only sales
    log. There is no distributed transaction here — stock is decremented
    first (the step that can legitimately fail, e.g. insufficient stock) and
    the sale is recorded only after that succeeds. If the Mongo insert
    itself fails, stock has already moved and this raises without rolling
    it back; a production system would reconcile that gap with an
    outbox/saga instead of a bare two-step call."""

    def __init__(self, products: IProductRepository, sales: ISaleRepository) -> None:
        self._products = products
        self._sales = sales
        self._logger = LoggerFactory.create_for_layer(
            service_name="clean-api-python", layer="application", component="RegisterSaleCommandHandler",
        )

    def handle(self, cmd: RegisterSaleCommand) -> str:
        if cmd.quantity <= 0:
            raise ValidationException(
                "quantity must be greater than 0",
                code=ErrorCodes.APP_VALIDATION_FAILURE,
            )

        product = self._products.find_by_id(cmd.product_id)
        if not product:
            self._logger.warning("Product not found", context={"id": cmd.product_id})
            raise ResourceNotFoundException(
                "Product not found", resource_type="Product", resource_id=cmd.product_id,
            )

        try:
            self._products.adjust_stock(cmd.product_id, -cmd.quantity)
        except InsufficientStockError:
            self._logger.warning(
                "Insufficient stock for sale",
                context={"product_id": cmd.product_id, "quantity": cmd.quantity},
            )
            raise ResourceConflictException(
                "insufficient stock for this sale",
                resource_type="Product",
                conflict_field="stock",
                conflict_value=str(cmd.quantity),
                code=ErrorCodes.APP_CONFLICT,
            )

        sale = Sale(product_id=cmd.product_id, quantity=cmd.quantity, unit_price=product.price)
        try:
            saved = self._sales.save(sale)
        except Exception as exc:
            # Stock was already decremented in Postgres — flagged here rather
            # than silently propagated, since this is exactly the
            # inconsistency window a saga/outbox pattern exists to close.
            db_exc = DatabaseException(
                "Sale record failed after stock was decremented - data inconsistency",
                operation="save_sale",
                table="sales",
                original_error=str(exc),
            )
            self._logger.log_kernel_exception(db_exc)
            raise db_exc from exc

        self._logger.info(
            "Sale registered",
            context={"sale_id": saved.id, "product_id": cmd.product_id, "quantity": cmd.quantity},
        )
        return saved.id
